Fix cursor movement to the beginning and end of words

gchp opened the pull request:
This fixes #35

gchp committed:
Fixed Cursor.EndWord skipping last word on a line

gchp committed:
Fix Cursor.PrevWord return value

kisielk commented on buffer/cursor.go:
```diff
@@ -280,9 +280,6 @@ func (c *Cursor) NextRuneFunc(f func(rune) bool) bool {
 			c.Boffset += rlen
 			r, rlen = c.RuneUnder()
 		}
-		if c.EOL() {
```
Won't this change keep it from moving on to the next line, since we'll hit the break below?

gchp replied:
Yes, actually. The tests must not have covered this, and it worked fine in my own testing. I've updated it now to check if EOL, and also to run the method passed in also. This should allow it to continue I think.

kisielk replied:
Hm, the logic in the function still seems to be too complicated, there's checks for EOL in three different places, and f(r) is checked twice. We should be able to do it with just one check of each.

gchp committed:
Refactor Cursor.NextRuneFunc to be more efficient

gchp commented:
@kisielk I've updated the `NextRuneFuc` method. I've changed it to use recursion, and I think it is much better. Let me know what you think.

---

Looking at it again, maybe it's a little verbose. This might be better:

``` go
// Move cursor forward until current rune satisfies condition f.
// Returns true if the move was successful, false if EOF reached.
func (c *Cursor) NextRuneFunc(f func(rune) bool) bool {
    r, rlen := c.RuneUnder()
    for !f(r) {
        if c.EOL() {
            next := c.NextLine()
            if !next {
                return false
            }
            c.Boffset = 0
        }
        c.Boffset += rlen
        r, rlen = c.RuneUnder()
    }
    return true
}
```

Though there is one failing test with this version. Let me know anyways.

kisielk commented:
Yeah that looks much better.

gchp commented:
@kisielk the version in the commit, or the one in the comment above?

kisielk commented:
the comment. Which test is failing?

gchp commented:
When I apply the change from the comment, this is the test output:

```
--- FAIL: TestNextWord (0.00 seconds)
    cursor_test.go:179: Bad cursor line at index 8 &{[32 114 101 116 117 114 110 32 48] 0xc210037f60 0xc210037f00} != &{[125] <nil> 0xc210037f30}
    cursor_test.go:182: Bad cursor position at index 8 9 != 0
```

It appears that `c.EOL()` is not detected properly, as it is moving one rune past the last rune in the line, instead of moving to the start of the next line.

kisielk commented:
Yeah, I was just discussing this with @dtcaciuc earlier today. The model for dealing with lines, end of lines, and cursor positions is a bit inconsistent. I think we have to rethink the idea from first principles and go over the stuff in the buffer package and test it to make sure it all behaves consistently. I will create a new issue for the criteria where we can discuss this.

gchp commented:
Cool. So I guess this PR should be put on hold then until that is resolved? I'll update with the example above, then fix the failing test when the cursor issue is resolved.

kisielk commented:
That's right. I think it's worth making sure everything is built on a solid foundation, otherwise when we try to fix cursor bugs in the future it will be a source of a lot of potential bugs in other parts of the code which rely on the wrong behaviour.

yarbelk commented:
I think that the cursor model is contributing to #44
